main.py

```diff
@@ -36,58 +37,72 @@ def main():
             batch=batch_number, state=poller_state))
         sierra_raw_data = sierra_client.execute_query(
             build_new_patrons_query(poller_state['creation_dt']))
-        sierra_df = pd.DataFrame(data=sierra_raw_data, dtype='string',
-                                 columns=['patron_id_plaintext', 'ptype_code',
-                                          'pcode3', 'patron_home_library_code',
-                                          'city', 'region', 'postal_code',
-                                          'address', 'circ_active_date_et',
-                                          'last_updated_date_et',
-                                          'deletion_date_et',
-                                          'creation_timestamp'])
+        unprocessed_sierra_df = pd.DataFrame(
+            data=sierra_raw_data, dtype='string',
+            columns=['patron_id_plaintext', 'ptype_code', 'pcode3',
+                     'patron_home_library_code', 'city', 'region',
+                     'postal_code', 'address', 'circ_active_date_et',
+                     'last_updated_date_et', 'deletion_date_et',
+                     'creation_timestamp'])
+
+        # Reduce the dataframe to only one row per patron_id, keeping the row
+        # with the lowest display_order and patron_record_address_type_id.
+        should_keep_mask = ~unprocessed_sierra_df.duplicated(
+            'patron_id_plaintext', keep='first')
+        processed_df = unprocessed_sierra_df[should_keep_mask].reset_index(
+            drop=True)
 
         # Get geoids from geocoder API and join with Sierra data
-        geoids = geocoder_client.get_geoids(sierra_df)
-        full_df = sierra_df.join(geoids)
+        geoids = geocoder_client.get_geoids(processed_df)
+        processed_df = processed_df.join(geoids)
 
         # Modify data to match what's expected by the PatronInfo Avro schema
-        full_df['ptype_code'] = pd.to_numeric(
-            sierra_df['ptype_code'], errors='coerce').astype('Int64')
-        full_df['pcode3'] = pd.to_numeric(
-            sierra_df['pcode3'], errors='coerce').astype('Int64')
-        full_df['creation_date_et'] = sierra_df['creation_timestamp'].apply(
-            _convert_dt_to_est_date)
+        processed_df['ptype_code'] = pd.to_numeric(
+            processed_df['ptype_code'], errors='coerce').astype('Int64')
+        processed_df['pcode3'] = pd.to_numeric(
+            processed_df['pcode3'], errors='coerce').astype('Int64')
+        processed_df['creation_date_et'] = processed_df[
+            'creation_timestamp'].apply(_convert_dt_to_est_date)
 
         # Obfuscate the patron ids and addresses using bcrypt
-        logger.info('Obfuscating ({}) patron ids'.format(len(full_df)))
-        full_df['patron_id'] = full_df['patron_id_plaintext'].apply(obfuscate)
+        logger.info('Obfuscating ({}) patron ids'.format(len(processed_df)))
+        with ThreadPoolExecutor() as executor:
```

Comment:
Should we set `max_workers` here (and below)? Looking at the docs since `3.8` it seems like it sets a reasonable default value but not sure about that.

Also: obfuscation is probably CPU bound, right? Should we use a process-based approach here?

Reply:
Yeah max_workers is set to min(32, (os.cpu_count() or 1) + 4) by default. I'm not sure how many CPUs an ECS instance has by default, so this seemed safe to me. I tested using both multithreading and multiprocessing locally and achieved pretty much the same results, but I'm planning on testing out both once I get the QA ECS cluster/task up and running.

Reply:
Sounds good. ECS size depends on the ec2 resources you allocate to it, but I think even a `t3.small` has one vCPU so that should be fine. On a tiny/micro it would be an issue, but we generally don't use those.